# 02 - Data Cleaning

## Objective

The objective of this notebook is to transform the raw Superstore dataset into a clean, analysis-ready dataset suitable for SQL analysis and Power BI dashboards.

## Tasks Performed

- Standardized column names
- Converted date columns to datetime
- Validated numeric fields
- Checked missing values
- Checked duplicate records
- Identified invalid records
- Performed outlier detection using IQR
- Exported cleaned dataset

In [55]:
# Import libraries
import pandas as pd
import numpy as np

In [56]:
# Load raw dataset
df = pd.read_csv("Superstore.csv", encoding="latin1")

In [57]:
# Display first five rows
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11-08-2016,11-11-2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11-08-2016,11-11-2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,06-12-2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10-11-2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10-11-2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [58]:
# Display Before cleaning column names
print(df.columns.tolist())

['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode', 'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State', 'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit']


In [59]:
# Standardize Column names
df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ","_")
    .str.replace("-","_")
)

In [60]:
# Display after cleaning column name
print(df.columns.tolist())

['row_id', 'order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_id', 'customer_name', 'segment', 'country', 'city', 'state', 'postal_code', 'region', 'product_id', 'category', 'sub_category', 'product_name', 'sales', 'quantity', 'discount', 'profit']


In [61]:
# Display sample date values
df[['order_date','ship_date']].head(10)

,order_date,ship_date
0,11-08-2016,11-11-2016
1,11-08-2016,11-11-2016
2,06-12-2016,6/16/2016
3,10-11-2015,10/18/2015
4,10-11-2015,10/18/2015
5,06-09-2014,6/14/2014
6,06-09-2014,6/14/2014
7,06-09-2014,6/14/2014
8,06-09-2014,6/14/2014
9,06-09-2014,6/14/2014


In [62]:
# convert data columns to datetime
df["order_date"] = pd.to_datetime(
    df["order_date"],
    format = "mixed",
    dayfirst = True
)

df["ship_date"] = pd.to_datetime(
    df["ship_date"],
    format = "mixed",
    dayfirst = True
)

In [63]:
df[["order_date", "ship_date"]].dtypes

order_date    datetime64[ns]
ship_date     datetime64[ns]
dtype: object

In [64]:
print(df["order_date"].isna().sum())
print(df["ship_date"].isna().sum())

0
0


In [65]:
df.head()

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,postal_code,region,product_id,category,sub_category,product_name,sales,quantity,discount,profit
0,1,CA-2016-152156,2016-08-11,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,2016-08-11,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,2016-12-06,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,2015-11-10,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,2015-11-10,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [66]:
# summary statistics for numeric columns
df[["sales", "quantity","discount","profit"]].describe()

,sales,quantity,discount,profit
count,9994.000000,9994.000000,9994.000000,9994.000000
mean,229.858001,3.789574,0.156203,28.656896
std,623.245101,2.225110,0.206452,234.260108
min,0.444000,1.000000,0.000000,-6599.978000
25%,17.280000,2.000000,0.000000,1.728750
50%,54.490000,3.000000,0.200000,8.666500
75%,209.940000,5.000000,0.200000,29.364000
max,22638.480000,14.000000,0.800000,8399.976000


In [67]:
# Orders with zero or nefative sales
invalid_sales = df[df["sales"] <= 0]

print(f"Invalid sales records: {len(invalid_sales)}")

Invalid sales records: 0


In [68]:
invalid_quantity = df[df["quantity"] <= 0]
print(f"Invalid Quantity Records: {len(invalid_quantity)}")

Invalid Quantity Records: 0


In [81]:
invalid_discount = df[ (df["discount"] < 0) | (df["discount"] > 1)]

print(f"Invalid Discount Records: {len(invalid_discount)}")

Invalid Discount Records: 0


In [70]:
loss_orders = df[df["profit"] < 0]
print(f"Loss Making Orders: {len(loss_orders)}")

Loss Making Orders: 1871


In [71]:
def detect_outliers_iqr(data,column):
        # Detect outliers in a numeric column using the IQP method.

    # Calculate Q1 and Q3
    Q1 = data[column].quantile(0.25)
    Q3 = data[column].quantile(0.75)

    # Calculate IQR
    IQR = Q3 - Q1

    # Calculate lower and upper bounds
    lower_bound = Q1 - (1.5 * IQR)
    upper_bound = Q3 + (1.5 * IQR)

    # Identify outliers
    outliers = data[ (data[column] < lower_bound) | (data[column] > upper_bound)]

    print(f"{column}")
    print(f"Lower Bound : {lower_bound : 2f}")
    print(f"Upper Bound : {upper_bound : 2f}")
    print(f"Outliers : {len(outliers)}")

    return outliers

In [72]:
sales_outliers = detect_outliers_iqr(df,"sales")

sales
Lower Bound : -271.710000
Upper Bound :  498.930000
Outliers : 1167


Outliers were detected using the Interquartile Range (IQR) method. A total of 1,167 sales transactions exceeded the upper bound of $498.93. These records were reviewed and retained because they represent legitimate high-value retail transactions rather than data entry errors. Removing them would distort revenue, customer, and product analyses.

In [73]:
profit_outliers = detect_outliers_iqr(df, "profit")

profit
Lower Bound : -39.724125
Upper Bound :  70.816875
Outliers : 1881


Using the IQR method, 1,881 profit outliers were identified.

These records were retained because they represent genuine business scenarios, including highly profitable orders and loss-making transactions. Retaining these observations preserves the integrity of profitability analysis and supports business decision-making.

In [78]:
# Save cleaned dataset
output_path = "data/processed/Superstore_Clean.csv"
df.to_csv(output_path, index=False)
print("✅ Cleaned dataset saved successfully!")

✅ Cleaned dataset saved successfully!


In [79]:
import os 
print(os.path.exists("data/processed/Superstore_Clean.csv"))

True


In [82]:
# ==========================================
# DATA VALIDATION SUMMARY REPORT
# ==========================================

validation_summary = {
    "Metric": [
        "Total Rows",
        "Total Columns",
        "Missing Values",
        "Duplicate Rows",
        "Invalid Sales Records",
        "Invalid Quantity Records",
        "Invalid Discount Records",
        "Loss Making Orders",
        "Sales Outliers",
        "Profit Outliers"
    ],
    
    "Value": [
        df.shape[0],
        df.shape[1],
        df.isnull().sum().sum(),
        df.duplicated().sum(),
        len(invalid_sales),
        len(invalid_quantity),
        len(invalid_discount),
        len(loss_orders),
        len(sales_outliers),
        len(profit_outliers)
    ]
}

validation_report = pd.DataFrame(validation_summary)

validation_report

,Metric,Value
0,Total Rows,9994
1,Total Columns,21
2,Missing Values,0
3,Duplicate Rows,0
4,Invalid Sales Records,0
5,Invalid Quantity Records,0
6,Invalid Discount Records,0
7,Loss Making Orders,1871
8,Sales Outliers,1167
9,Profit Outliers,1881
